src/renderers/shared/fiber/ReactChildFiber.js

```diff
@@ -43,6 +43,38 @@ if (__DEV__) {
   var getComponentName = require('getComponentName');
   var warning = require('fbjs/lib/warning');
   var didWarnAboutMaps = false;
+  /**
+   * Warn if there's no key explicitly set on dynamic arrays of children or
+   * object keys are not valid. This allows us to keep track of children between
+   * updates.
+   */
+  var ownerHasKeyUseWarning = {};
+
+  var warnForMissingKey = (child: ReactElement) => {
+    if (!child._store || child._store.validated || child.key != null) {
+      return;
+    }
+    child._store.validated = true;
+
+    var memoizer = ownerHasKeyUseWarning.uniqueKey ||
+      (ownerHasKeyUseWarning.uniqueKey = {});
```

Comment:
It was added here - https://github.com/facebook/react/commit/47e49ae8b7e41b566db4cabb47ee2f31e5b9bf0b and looks like the author was just copying the type of logic that was used before, when we had two nested objects for the two types of key warnings we threw. So I am removing the nesting.